[AG-38479] Updated userscript development section in extensions article (#632)
AdguardTeam/KnowledgeBase: docs/general/extensions.md

- [`GM.setValue`, `GM_setValue`](https://wiki.greasespot.net/GM.setValue)
- [`GM.getValue`, `GM_getValue`](https://wiki.greasespot.net/GM.getValue)
- [`GM.listValues`, `GM_listValues`](https://wiki.greasespot.net/GM.listValues)
- [`GM.deleteValue`, `GM_deleteValue`](https://wiki.greasespot.net/GM.deleteValue)
- [`GM.getResourceUrl`, `GM_getResourceURL`](https://wiki.greasespot.net/GM.getResourceUrl)
- [`GM.setClipboard`, `GM_setClipboard`](https://wiki.greasespot.net/GM.setClipboard)
- [`GM.xmlHttpRequest`, `GM_xmlhttpRequest`](https://wiki.greasespot.net/GM.xmlHttpRequest)
- [`GM.openInTab`, `GM_openInTab`](https://wiki.greasespot.net/GM.openInTab)
- [`GM.notification`](https://wiki.greasespot.net/GM.notification)
- [`unsafeWindow`](https://wiki.greasespot.net/UnsafeWindow)
- [`GM_getResourceText`](https://www.tampermonkey.net/documentation.php#api:GM_getResourceText)
- [`GM_addStyle`](https://www.tampermonkey.net/documentation.php#api:GM_addStyle)
- [`GM_log`](https://www.tampermonkey.net/documentation.php#api:GM_log)
- [`GM.addElement`, `GM_addElement`](https://www.tampermonkey.net/documentation.php#api:GM_addElement)

> [Here](https://wiki.greasespot.net/Greasemonkey_Manual:API) you can find more information about Greasemonkey API.

Reviewer: ```suggestion
You can find more information about Greasemonkey API in [its manual](https://wiki.greasespot.net/Greasemonkey_Manual:API) .
```

// @grant           GM_info
// @grant           GM_openInTab
// @grant           GM_registerMenuCommand
// @grant           GM_addElement
// @run-at          document-start
// ==/UserScript==
!function(){(
    console.log("I am loaded!");
)}();
```

> #### Trusted Types Policy API

Reviewer: нашла TrustedTypePolicy (https://developer.mozilla.org/en-US/docs/Web/API/TrustedTypePolicy) и Trusted Types API (https://developer.mozilla.org/en-US/docs/Web/API/Trusted_Types_API)

Author: Переименовал в `Trusted Types API`

> AdGuard provides an instance of `PolicyApi` class that allows you to manage Trusted Types in your userscripts.

Reviewer: ```suggestion
AdGuard provides an instance of the `PolicyApi` class that allows you to manage Trusted Types in your userscripts.
```

> You can access instance of this class by using `ADG_policyApi` variable in your userscript.

Reviewer: ```suggestion
You can access the instance of this class by using the `ADG_policyApi` variable in your userscript.
```

##### Properties

> - `name: string` - Name of the policy (Default is `"AGPolicy"`).

Reviewer: ```suggestion
- `name: string` — a name of the policy (Default is `"AGPolicy"`).
```

> - `isSupported: boolean` - Flag that indicates is Trusted Types API supported or not in the current browser.

Reviewer: ```suggestion
- `isSupported: boolean` — a flag indicating whether or not the Trusted Types API is supported by the current browser.

##### Polyfilled methods

> - [`ADG_policyApi.createHTML`]. If not supported returns `input: string` back.

Reviewer: ```suggestion
- [`ADG_policyApi.createHTML`]. If not supported, returns `input: string`.
```

Reviewer: ниже аналогично

- [`ADG_policyApi.createScript`]. If not supported returns `input: string` back.
- [`ADG_policyApi.createScriptURL`]. If not supported returns `input: string` back.
- [`ADG_policyApi.getAttributeType`]. If not supported returns `null`.
- [`ADG_policyApi.getPropertyType`]. If not supported returns `null`.
- [`ADG_policyApi.isHTML`]. If not supported returns `false`.
- [`ADG_policyApi.isScript`]. If not supported returns `false`.
- [`ADG_policyApi.isScriptURL`]. If not supported returns `false`.

> [`ADG_policyApi.createHTML`]: https://developer.mozilla.org/en-US/docs/Web/API/TrustedTypePolicy/createHTML

Reviewer: ```suggestion
[`ADG_policyApi.createHTML`](https://developer.mozilla.org/en-US/docs/Web/API/TrustedTypePolicy/createHTML)
```

Reviewer: мб в таком формате?

[`ADG_policyApi.createScript`]: https://developer.mozilla.org/en-US/docs/Web/API/TrustedTypePolicy/createScript
[`ADG_policyApi.createScriptURL`]: https://developer.mozilla.org/en-US/docs/Web/API/TrustedTypePolicy/createScriptURL
[`ADG_policyApi.getAttributeType`]: https://developer.mozilla.org/en-US/docs/Web/API/TrustedTypePolicyFactory/getAttributeType
[`ADG_policyApi.getPropertyType`]: https://developer.mozilla.org/en-US/docs/Web/API/TrustedTypePolicyFactory/getPropertyType
[`ADG_policyApi.isHTML`]: https://developer.mozilla.org/en-US/docs/Web/API/TrustedTypePolicyFactory/isHTML
[`ADG_policyApi.isScript`]: https://developer.mozilla.org/en-US/docs/Web/API/TrustedTypePolicyFactory/isScript
[`ADG_policyApi.isScriptURL`]: https://developer.mozilla.org/en-US/docs/Web/API/TrustedTypePolicyFactory/isScriptURL

##### Additional Types

```typescript
/**

>  * Enum representation of return values of `getAttributeType` and

Reviewer: ```suggestion
 * Enum representation of the return values of the `getAttributeType` and
```

>  * `getPropertyType` methods of native Trusted Types API.

Reviewer: ```suggestion
 * `getPropertyType` methods of the native Trusted Types API.
```

 *
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/TrustedTypePolicyFactory/getAttributeType}
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/TrustedTypePolicyFactory/getPropertyType}
 */
enum TrustedType {
    HTML = 'TrustedHTML',
    Script = 'TrustedScript',
    ScriptURL = 'TrustedScriptURL',
}

// You can access it like that inside of userscript
ADG_TrustedType.HTML // "TrustedHTML"

/**

>  * Isomorphic trusted value type, if browser supports Trusted Types API it will be one of the

Reviewer: ```suggestion
 * Isomorphic trusted value type. If a browser supports the Trusted Types API, it will be one of the enum Trusted Types
```

>  * `TrustedHTML`, `TrustedScript` or `TrustedScriptURL`, otherwise it will be regular `string`.

Reviewer: ```suggestion
 * (`TrustedHTML`, `TrustedScript` or `TrustedScriptURL`); otherwise, it will be regular `string`.
```

 *
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/TrustedHTML}
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/TrustedScript}
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/TrustedScriptURL}
 */
type TrustedValue = string | TrustedHTML | TrustedScript | TrustedScriptURL;
```

##### Additional methods

```typescript
/**

>  * Creates Trusted Type depending on `type`:

Reviewer: ```suggestion
 * Creates a Trusted Type depending on `type`:
```

 * - `TrustedHTML`
 * - `TrustedScript`
 * - `TrustedScriptURL`

>  * - or returns back `value` if none of them applicable.

Reviewer: ```suggestion
 * - or returns `value` if none of them is applicable.
```

 *

>  * @param type Trusted Type.

Reviewer: ```suggestion
 * @param type — Trusted Type.
```

Reviewer: ниже аналогично

Author: Зарезолвил похожие саджешн тк это jsdoc и тут не используется "-" и не начинаем с артикля

>  * @param value Value from which creates Trusted Type.

Reviewer: ```suggestion
 * @param value — a value from which a Trusted Type is created.
```

 * @param createArgs Additional arguments to be passed to the function represented by `TrustedTypePolicy`.
 * @returns Created value.
 */
function create(
    type: TrustedType,
    value: string,
    ...createArgs: unknown[]
): TrustedValue

// Example: Creates TrustedHTML
const trustedHTML = ADG_policyApi.create(ADG_TrustedType.HTML, '<div></div>');

/**
 * Converts `value` of `attribute` into one of the Trusted Types:
 * - `TrustedHTML`
 * - `TrustedScript`
 * - `TrustedScriptURL`

>  * - or returns back `value` if none of them applicable.

Reviewer: ```suggestion
 * - or returns `value` if none of them applicable.
```

>  * - or returns back `value` if none of them applicable.

Reviewer: ```suggestion
 * - or returns `value` if none of them is applicable.
```

 *
 * @param tagName Name of an HTML tag.
 * @param attribute Attribute.
 * @param value Value of attribute that needs to be converted.
 * @param elementNS Element namespace, if empty defaults to the HTML namespace.
 * @param attrNS Attribute namespace, if empty defaults to null.
 * @param createArgs Additional arguments to be passed to the function represented by `TrustedTypePolicy`.
 * @returns Converted value.
 */
function convertAttributeToTrusted(
    tagName: string,
    attribute: string,
    value: string,
    elementNS?: string,
    attrNS?: string,
    ...createArgs: unknown[]
): TrustedValue

// Example: Converts to TrustedScriptURL
const trustedScriptURL = ADG_policyApi.convertAttributeToTrusted("script", "src", 'SOME_URL');
scriptElement.setAttribute("src", trustedScriptURL);

/**
 * Converts `value` of `property` into one of the Trusted Types:
 * - `TrustedHTML`
 * - `TrustedScript`
 * - `TrustedScriptURL`
 * - or returns back `value` if none of them applicable.
 *
 * @param tagName Name of an HTML tag.

>  * @param property Property.

Reviewer: ```suggestion
 * @param property — a property.
```

>  * @param value Value or property.

Reviewer: ```suggestion
 * @param value — a value or a property.
```

>  * @param elementNS Element namespace, if empty defaults to the HTML namespace.

Reviewer: ```suggestion
 * @param elementNS — an element namespace. If empty, defaults to the HTML namespace.
```

>  * @param createArgs Additional arguments to be passed to the function represented by `TrustedTypePolicy`.

Reviewer: ```suggestion
 * @param createArgs — additional arguments to be passed to the function represented by `TrustedTypePolicy`.
```

>  * @returns Converted value.

Reviewer: ```suggestion
 * @returns — a converted value.
```

 * - `TrustedHTML`
 * - `TrustedScript`
 * - `TrustedScriptURL`
 * - or returns `value` if none of them is applicable.
 *
 * @param tagName Name of an HTML tag.
 * @param attribute Attribute.

>  * @param value Value of an attribute to be converted.

Author: Это документация внутри кода, тут думаю "-" и артикли не нужны.
Все таски ниже резолвлю тоже

 * - `TrustedHTML`
 * - `TrustedScript`
 * - `TrustedScriptURL`
 * - or returns `value` if none of them is applicable.
 *
 * @param tagName Name of an HTML tag.
 * @param property Property.
 * @param value Value of a property to be converted.
 * @param elementNS Element namespace. If empty, defaults to the HTML namespace.
 * @param createArgs Additional arguments to be passed to the function represented by `TrustedTypePolicy`.

>  * @returns Converted value.

Reviewer: ```suggestion
 * @returns — a converted value.
```
